### PR TITLE
[red-knot] Infer float and complex literal expressions

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR implements type inference for float and complex literal expressions.

## Test Plan

Add test cases for both types.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 1]
```diff
@@ -1639,12 +1639,23 @@ mod tests {
...
+            c = 1.45
+            d = 2j
+            ",
+        )?;
 
-        assert_public_ty(&db, "src/a.py", "x", "Literal[1]");
+        assert_public_ty(&db, "src/a.py", "a", "Literal[1]");
+        assert_public_ty(&db, "src/a.py", "b", "Literal[int]");
+        assert_public_ty(&db, "src/a.py", "c", "Literal[float]");
+        assert_public_ty(&db, "src/a.py", "d", "Literal[complex]");
```
[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/red_knot_python_semantic/src/types/infer.rs:1658]: Yes, I think this is currently a limitation and must be resolved separately as is the case for other literal types as well (`Literal[set]`, `Literal[tuple]`, etc.).

[the remark this answers is not in the document]
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types/infer.rs:1658]: That seems quite an important limitation for us to resolve; I'd be inclined to defer making changes to `infer_number_literal_expression()` until it is resolved, in that case. I'm not sure what the use is of adding type inference for `float` and `complex` expressions if we infer the wrong type entirely ;-)

[the remark this answers is not in the document]
Comment (dhruvmanila) [reply crates/red_knot_python_semantic/src/types/infer.rs:1658]: Yeah, I think that makes sense. I'll put this in draft for the time being.

**Review by carljm — APPROVED** [review]
Yeah, sorry, this was my oversight in a previous PR. Rebasing on https://github.com/astral-sh/ruff/pull/12695 should allow fixing this; LGTM with that change.

**Pull request merged**